### DISCUSSION TITLE
Which are the valid namespace methods and events?

### DISCUSSION THREAD

**Discussion by moltam89** [WalletConnect/walletconnect-monorepo#2318] [Developer Support]
Hi,

I'm building V2 support for PunkWallet.io and I'm struggling to understand  which are the necessary methods and events.

I tried to use the [namespaces builder util ](https://docs.walletconnect.com/2.0/web/web3wallet/wallet-usage#-namespaces-builder-util) with methods **["eth_sendTransaction", "eth_signTransaction", "eth_sign", "personal_sign", "eth_signTypedData"]**, but this won't work with https://claim.lens.xyz/ , where the required namespace contains extra methods :**["eth_sendRawTransaction","eth_sign","eth_signTypedData_v3","eth_signTypedData_v4","wallet_switchEthereumChain","wallet_addEthereumChain"]**

I can see that in [the web examples repo](https://github.com/WalletConnect/web-examples/blob/c6e9c2d6149f05617f0125dccc3ac04a062540c8/wallets/react-web3wallet/src/views/SessionProposalModal.tsx#L60-L82) basically all methods are approved, but it feels fishy, because not all of these methods are actually implemented.
Which are the methods a wallet should implement?

Also I'm a bit confused about the events, in the examples I can see **events: ["accountsChanged", "chainChanged"]**, but the example wallets doesn't seem to listen to these events.

Thanks,
Tamas

**Comment by finessevanes** [comment]
Here are the docs over namespaces: https://docs.walletconnect.com/2.0/specs/clients/sign/namespaces

LMK if this unblocks you!

**Reply by moltam89** [reply]
Thanks for the reply!

Yeah, I saw this doc, but I was looking for specified list what a wallet should implement.
I use the [namespaces builder util](https://github.com/moltam89/punk-wallet/blob/50cc7bf45d29ffc892211c199716ac10c7461529/packages/react-app/src/helpers/WalletConnectV2Helper.js#L52-L53), with the methods and events required by the https://react-app.walletconnect.com/ example Dapp, which works perfectly fine.

This also works with https://www.lens.xyz/, however it is not enough for https://claim.lens.xyz/, because they require more methods.
It would be nice to have a convention which methods/events should be implemented. (Are there more V2 Dapps?)

If I understand it correctly, the example wallets just [approve](https://github.com/WalletConnect/web-examples/blob/c6e9c2d6149f05617f0125dccc3ac04a062540c8/wallets/react-web3wallet/src/views/SessionProposalModal.tsx#L60-L82) all the methods/events which are requested, which can solve the connection problem, but later the Dapp can require unimplemented actions.

Let me know if I misunderstood something.

**Reply by finessevanes** [reply]
>the example wallets just [approve](https://github.com/WalletConnect/web-examples/blob/c6e9c2d6149f05617f0125dccc3ac04a062540c8/wallets/react-web3wallet/src/views/SessionProposalModal.tsx#L60-L82) all the methods/events which are requested, which can solve the connection problem, but later the Dapp can require unimplemented actions.

Half and half.

>the example wallets just [approve](https://github.com/WalletConnect/web-examples/blob/c6e9c2d6149f05617f0125dccc3ac04a062540c8/wallets/react-web3wallet/src/views/SessionProposalModal.tsx#L60-L82) all the methods/events which are requested

This is accurate.

> later the Dapp can require unimplemented actions

The dapp will send both required and optional namespaces in the session proposal. Later, they can't require unimplemented actions.


Thank you for your suggestion. I've passed it along with our team! We're actually working on a PR to update the namespace doc with should be merged tomorrow.

**Comment by migalv** [comment]
I'm trying to implement the Wallet Connect v2 into my Flutter wallet app. There is very few documentation on how to do this. And the current documentation on Namespaces does not answer my questions.

I'm also having a lot of trouble understanding what events and methods I should add to my Session Namespace. I would like my wallet to work with most of the dApps that are out there.

I'm not very blockchain savvy so it's hard to keep track of all the new concepts and things. Could you point me into the right direction?
